### PR TITLE
Closing files and streams consume them

### PR DESCRIPTION
**Pull request by matklad**
### What does this PR do?

Change close from `&mut self` to `self`

### Motivation

Splitting signature-change part of #92.

### Checklist

[] I have added unit tests to the code I am submitting
[] My unit tests cover both failure and success scenarios
[] If applicable, I have discussed my architecture
[x] The new code I am adding is formatted using `rustfmt`

### CODE REVIEW

File: scipio/src/io/dma_file.rs [hunk 1]
```diff
@@ -517,12 +518,21 @@ impl DmaFile {
...
 
     /// Closes this DMA file.
-    pub async fn close(&mut self) -> io::Result<()> {
+    pub async fn close(mut self) -> io::Result<()> {
         let source = Reactor::get().close(self.as_raw_fd());
         enhanced_try!(source.collect_rw().await, "Closing", self)?;
         self.file = unsafe { std::fs::File::from_raw_fd(-1) };
         Ok(())
     }
+    pub(crate) async fn close_rc(self: Rc<DmaFile>) -> io::Result<()> {
```
Comment (matklad) [inline scipio/src/io/dma_file.rs:527]: Note that  `Rc<Self>` works with method call syntax.

File: scipio/src/io/file_stream.rs [hunk 2]
```diff
@@ -280,35 +280,15 @@ pub struct StreamReader {
 
 macro_rules! collect_error {
     ( $state:expr, $res:expr ) => {{
-        let mut state = $state.borrow_mut();
         if let Err(x) = $res {
-            state.error = Some(x.into());
+            $state.borrow_mut().error = Some(x.into());
```
Comment (matklad) [inline scipio/src/io/file_stream.rs:284]: microoptimization -- borrow only if there's an error!

File: scipio/src/io/file_stream.rs [hunk 3]
```diff
@@ -1392,15 +1374,6 @@ mod test {
         }
     }
 
-    file_stream_read_test!(read_close_twice, path, _k, file, _file_size: 131072, {
-        let mut reader = StreamReaderBuilder::new(file)
-            .with_buffer_size(1024)
-            .build();
-
-        reader.close().await.unwrap();
-        expect_specific_error(reader.close().await, "\"Bad file descriptor (os error 9)\"");
```
Comment (matklad) [inline scipio/src/io/file_stream.rs:1401]: This now doesn't compile -- one just can't close the file twice, it's gone after the first try.

**Comment by matklad** [comment]
Rebased! (are rebases on top of master allowed, or should I use merge workflow?)

**Comment by glommer** [comment]
I prefer rebases to keep the history clean, but I will usually adopt a "whatever works for you" approach until the PR is approved and ready to merge. That's just because sometimes rebasing is a pain and it's best to inflict the pain only once

But if rebasing is easy and/or is your personal preference anyway go for it.

Is this ready for review?

File: scipio/src/io/file_stream.rs (hunk 2 above)
Comment (glommer) [reply scipio/src/io/file_stream.rs:284]: agree, good change.

File: scipio/src/io/dma_file.rs (hunk 1 above)
Comment (glommer) [reply scipio/src/io/dma_file.rs:463]: this is indeed better.

All good and no objections. Pushing the button on this, thanks!

**Review by glommer — APPROVED** [review]

**Pull request merged**